working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,924 @@
+# Patterns
+
+**NOTE: This document is still *very* rough. Note the many many TODOs. Putting
+it out here now to start getting language team feedback on direction.**
+
+**TODO: Show examples of common interesting use cases.**
+
+If you want to get straight to the proposal, feel free to skip the next big
+section. If you want to understand *why* the proposal makes the choices it does,
+read on.
+
+## Design Challenges
+
+One cannot simply amputate the pattern matching syntax and semantics from a
+functional language like SML and graft it directly onto an object-oriented
+language like Dart. SML's type system is based on algebraic datatypes and has no
+notion of subtyping.
+
+The impedence mismatch between existing functional-style pattern syntax and
+Dart causes (at least) two challenges:
+
+### Challenge 1: Bare identifiers
+
+Consider this hypothetical Dart program:
+
+```dart
+const a = 123;
+
+main() {
+  match ([345]) {
+    case [a] => print(a);
```

Comment:
Likewise, are we matching a one-item `List` containing one element?

Reply:
Yes. `[a]` is a list destructuring pattern containing a single subpattern `a` that's a variable pattern.